neonvm-runner/cmd/main.go

```diff
@@ -669,6 +673,85 @@ func main() {
 	}
 }
 
+type NetworkMonitoringMetrics struct {
+	IngressBytes, EgressBytes       prometheus.Counter
+	Errors                          *prometheus.CounterVec
+	IngressBytesRaw, EgressBytesRaw uint64 // Absolute values to calc increments for Counters
+}
+
+func NewMonitoringMetrics(reg *prometheus.Registry) *NetworkMonitoringMetrics {
+	m := &NetworkMonitoringMetrics{
+		IngressBytes: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "runner_vm_ingress_bytes",
+				Help: "Number of bytes received by the VM from the open internet",
+			},
+		),
+		EgressBytes: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "runner_vm_egress_bytes",
+				Help: "Number of bytes sent by the VM to the open internet",
+			},
+		),
+		IngressBytesRaw: 0,
+		EgressBytesRaw:  0,
+		Errors: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "runner_vm_network_fetch_errors_total",
+				Help: "Number of errors while fetching network monitoring data",
+			},
+			[]string{"cause"},
+		),
+	}
+	reg.MustRegister(m.IngressBytes)
+	reg.MustRegister(m.EgressBytes)
+	reg.MustRegister(m.Errors)
+	return m
+}
+
+func getNetworkBytesCounter(iptables *iptables.IPTables, chain string) (uint64, error) {
+	cnt := uint64(0)
+	rules, err := iptables.Stats("filter", chain)
+	if err != nil {
+		return cnt, err
+	}
+	for _, rawStat := range rules {
+		stat, err := iptables.ParseStat(rawStat)
+		if err != nil {
+			return cnt, err
+		}
+		if stat.Protocol == "6" { // tcp
+			cnt += stat.Bytes
+		}
+	}
+	return cnt, nil
+}
+
+func (m *NetworkMonitoringMetrics) increment() {
+	// Rules configured at github.com/neondatabase/cloud/blob/main/compute-init/compute-init.sh#L98
+	iptables, err := iptables.New()
+	if err != nil {
+		m.Errors.WithLabelValues(util.RootError(err).Error()).Inc()
+		return
+	}
+
+	if ingress, err := getNetworkBytesCounter(iptables, "INPUT"); err != nil {
+		m.Errors.WithLabelValues(util.RootError(err).Error()).Inc()
+		return
+	} else {
+		m.IngressBytes.Add(float64(ingress - m.IngressBytesRaw))
+		m.IngressBytesRaw = ingress
+	}
+
+	if egress, err := getNetworkBytesCounter(iptables, "OUTPUT"); err != nil {
+		m.Errors.WithLabelValues(util.RootError(err).Error()).Inc()
+		return
+	} else {
+		m.EgressBytes.Add(float64(egress - m.EgressBytesRaw))
+		m.EgressBytesRaw = egress
```

Comment:
IIUC, here and above, we do this `Add(previous value - current value)` — am I understanding this correctly, and if so, can we just `Set(current value)` instead?

And further: If we _do_ change it that way, then we should probably rename this method from `increment()` to something that more closely matches the content of what it's doing: setting the metric values based on the current iptables counts.

Reply:
No, because this is a Counter, not a Gauge, thus we are incrementing

Reply:
Should we make it a gauge then, to make this simpler?

Reply:
Technically it's a Counter as bytes can't go down (only if iptable rules are reset which doesn't happend with us) so I'd better expose the Counter predicate